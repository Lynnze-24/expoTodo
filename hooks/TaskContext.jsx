import React, { createContext, useContext, useState,useEffect} from 'react'
import uuid from 'react-native-uuid';
import AsyncStorage from '@react-native-async-storage/async-storage';

const TaskContext = createContext();


export function useTaskContext(){
    return useContext(TaskContext)
}

export function TaskProvider({children,mock,val}) {
    // mock and val are for testing
    const initial = mock?mock : [];
    const [task, setTask] = useState(initial);
    const fetchTasks = mock||val? ()=>{return}:async() => {
        let localTasks= await AsyncStorage.getItem('TaskAHL') || null;
        if(localTasks){
            setTask(JSON.parse(localTasks));
        }
    }// second function is for testing 

    
    useEffect(()=> {
        fetchTasks()
    
    }, [])
    
    const addToTask = async(item)=>{
        let newTask = [...task,item]
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    const removeFromTask = async(id)=>{
        let newTask = task.filter(x=> x.id!== id);
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }

    const toggleCheck = async(id) =>{
        let newTask = [...task];
       let changedItem = newTask.find(x=> x.id=== id);
       changedItem.isChecked = !changedItem.isChecked;
       setTask(newTask);
       AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    
    const value = val || {task,addToTask,removeFromTask,toggleCheck};
    
    return (
        <TaskContext.Provider value={value}>
        {children}
      </TaskContext.Provider>
    )
}