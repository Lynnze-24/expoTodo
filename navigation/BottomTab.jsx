import { createMaterialBottomTabNavigator } from '@react-navigation/material-bottom-tabs';
import All from '../screens/todo/All';
import Complete from '../screens/todo/Complete';
import Incomplete from '../screens/todo/Incomplete';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import { FontAwesome } from '@expo/vector-icons'; 
import theme from '../theme/Theme';

const Tab = createMaterialBottomTabNavigator();

export default function BottomTab() {
  
  return (
    <Tab.Navigator 
        initialRouteName="All"
        activeColor={theme.secondary}
        inactiveColor={theme.disabled}
        shifting='true'
        
        barStyle={{ backgroundColor: theme.primary}}>
      <Tab.Screen 
        name="All" 
        component={All} 
        
        options={{
          tabBarLabel: 'All',
          tabBarTestID:'AllBtn',
          tabBarIcon: ({ color,focused }) => (
            <MaterialCommunityIcons  name="home" size={26} color={color} />)}}
            />
      <Tab.Screen 
        name="Completed" 
        component={Complete} 
        options={{
          tabBarLabel: 'Completed',
          tabBarTestID:'CompleteBtn',
          tabBarIcon: ({ color }) => (
            <MaterialCommunityIcons  name="playlist-check" size={27} color={color}  />
          ),}}/>
      <Tab.Screen
        name="Imcomplete" 
        component={Incomplete} 
        options={{
          tabBarLabel: 'Incomplete',
          tabBarTestID:'IncompleteBtn',
          tabBarIcon: ({ color,focused }) => (
            <FontAwesome  name="tasks" size={16} color={color}  style={{marginTop:4}}/>
          ),}}/>
    </Tab.Navigator>
  );
}

