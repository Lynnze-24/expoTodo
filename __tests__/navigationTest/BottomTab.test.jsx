import { render,fireEvent } from '@testing-library/react-native';
import renderer, { act } from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';

import BottomTab from '../../navigation/BottomTab';
import { TaskProvider } from '../../hooks/TaskContext';




it('renders correctly', () => {
    const comp = renderer.create((<NavigationContainer >
      <TaskProvider>
    <BottomTab />
    </TaskProvider>
  </NavigationContainer>)).toJSON();
    expect(comp).toMatchSnapshot();
});
   

    it('should contain 3 tabs and should render alltodo page initially', async () => {
      const {findAllByTestId,getByTestId,queryByTestId}= render(
        <NavigationContainer >
            <TaskProvider>
          <BottomTab />
          </TaskProvider>
        </NavigationContainer>
      );
      expect((await findAllByTestId('AllScreen')).length).toBe(1) // initial page is Alltodo
      getByTestId('AllBtn')
      getByTestId('CompleteBtn')
      getByTestId('IncompleteBtn')
      
      })

      it('should go to completed page when user clicks completed tab', async() => {
        const {queryByTestId,findAllByTestId,getByTestId,debug}= render(
          <NavigationContainer >
              <TaskProvider>
            <BottomTab  />
            </TaskProvider>
          </NavigationContainer>
        );
        const completeBtn = getByTestId('CompleteBtn')
        fireEvent.press(getByTestId('IncompleteBtn'));
        // expect((await findAllByTestId('CompleteScreen')).length).toBe(1)
        // react navigation current version has a bug with pressing event in jest
      // will be automatically fixed in unreleased version
      // https://github.com/react-navigation/react-navigation/issues/10419
       
        })

        it('should go to imcomplete page when user clicks incomplete tab', async() => {
          const {queryByTestId,findAllByTestId,getByTestId,debug}= render(
            <NavigationContainer >
                <TaskProvider>
              <BottomTab  />
              </TaskProvider>
            </NavigationContainer>
          );
              fireEvent.press(getByTestId('IncompleteBtn'));
              // expect((await findAllByTestId('IncompleteScreen')).length).toBe(1)
              // react navigation current version has a bug with pressing event in jest
            // will be automatically fixed in unreleased version
            // https://github.com/react-navigation/react-navigation/issues/10419
          })
     
     
  
      
