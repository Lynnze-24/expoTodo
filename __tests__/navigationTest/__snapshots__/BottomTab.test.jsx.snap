// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
          },
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        collapsable={false}
        importantForAccessibility="auto"
        pointerEvents="auto"
        removeClippedSubviews={false}
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "display": "flex",
            },
          ]
        }
        testID="RouteScreen: 0"
      >
        <View
          collapsable={false}
          style={
            Object {
              "flex": 1,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fffeee",
                "flex": 1,
                "paddingHorizontal": 20,
              }
            }
            testID="AllScreen"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
              testID="emptyTask"
            >
              <Text>
                There is no task
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "#120311",
          "bottom": 0,
          "elevation": 4,
          "left": 0,
          "position": null,
          "right": 0,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 4,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(18, 3, 17, 1)",
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="tablist"
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "marginBottom": 0,
                "marginHorizontal": 0,
                "maxWidth": 504,
              },
            ]
          }
        >
          <View
            collapsable={false}
            pointerEvents="none"
            style={
              Object {
                "backgroundColor": undefined,
                "borderRadius": 0,
                "height": 0,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 28,
                "transform": Array [
                  Object {
                    "scale": 0.008,
                  },
                ],
                "width": 0,
              }
            }
          />
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "selected": true,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "paddingVertical": 6,
              }
            }
            testID="AllBtn"
          >
            <View
              pointerEvents="none"
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignSelf": "center",
                    "height": 24,
                    "marginHorizontal": 12,
                    "marginTop": 2,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "left": 0,
                        "position": "absolute",
                        "top": -2,
                      },
                      Object {
                        "right": -2,
                      },
                    ]
                  }
                >
                  <Text
                    collapsable={false}
                    numberOfLines={1}
                    style={
                      Object {
                        "alignSelf": "flex-end",
                        "backgroundColor": "#f50057",
                        "borderRadius": 8,
                        "color": "#ffffff",
                        "fontFamily": "System",
                        "fontSize": 8,
                        "fontWeight": "400",
                        "height": 16,
                        "lineHeight": 16,
                        "minWidth": 16,
                        "opacity": 0,
                        "overflow": "hidden",
                        "paddingHorizontal": 4,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  />
                </View>
              </View>
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 16,
                    "paddingBottom": 2,
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "fontSize": 12,
                            "height": 56,
                            "textAlign": "center",
                          },
                          Object {
                            "color": "#f0edf6",
                          },
                        ],
                      ]
                    }
                  >
                    All
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "selected": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "paddingVertical": 6,
              }
            }
            testID="CompleteBtn"
          >
            <View
              pointerEvents="none"
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignSelf": "center",
                    "height": 24,
                    "marginHorizontal": 12,
                    "marginTop": 2,
                    "transform": Array [
                      Object {
                        "translateY": 7,
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "left": 0,
                        "position": "absolute",
                        "top": -2,
                      },
                      Object {
                        "right": -2,
                      },
                    ]
                  }
                >
                  <Text
                    collapsable={false}
                    numberOfLines={1}
                    style={
                      Object {
                        "alignSelf": "flex-end",
                        "backgroundColor": "#f50057",
                        "borderRadius": 8,
                        "color": "#ffffff",
                        "fontFamily": "System",
                        "fontSize": 8,
                        "fontWeight": "400",
                        "height": 16,
                        "lineHeight": 16,
                        "minWidth": 16,
                        "opacity": 0,
                        "overflow": "hidden",
                        "paddingHorizontal": 4,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  />
                </View>
              </View>
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 16,
                    "paddingBottom": 2,
                    "transform": Array [
                      Object {
                        "scale": 0.5,
                      },
                    ],
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "fontSize": 12,
                            "height": 56,
                            "textAlign": "center",
                          },
                          Object {
                            "color": "#f0edf6",
                          },
                        ],
                      ]
                    }
                  >
                    Completed
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "selected": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "paddingVertical": 6,
              }
            }
            testID="IncompleteBtn"
          >
            <View
              pointerEvents="none"
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignSelf": "center",
                    "height": 24,
                    "marginHorizontal": 12,
                    "marginTop": 2,
                    "transform": Array [
                      Object {
                        "translateY": 7,
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  collapsable={false}
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "left": 0,
                        "position": "absolute",
                        "top": -2,
                      },
                      Object {
                        "right": -2,
                      },
                    ]
                  }
                >
                  <Text
                    collapsable={false}
                    numberOfLines={1}
                    style={
                      Object {
                        "alignSelf": "flex-end",
                        "backgroundColor": "#f50057",
                        "borderRadius": 8,
                        "color": "#ffffff",
                        "fontFamily": "System",
                        "fontSize": 8,
                        "fontWeight": "400",
                        "height": 16,
                        "lineHeight": 16,
                        "minWidth": 16,
                        "opacity": 0,
                        "overflow": "hidden",
                        "paddingHorizontal": 4,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  />
                </View>
              </View>
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 16,
                    "paddingBottom": 2,
                    "transform": Array [
                      Object {
                        "scale": 0.5,
                      },
                    ],
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "fontSize": 12,
                            "height": 56,
                            "textAlign": "center",
                          },
                          Object {
                            "color": "#f0edf6",
                          },
                        ],
                      ]
                    }
                  >
                    Incomplete
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
